### PR TITLE
Update CI and workflows for Python 3.12 support

### PR DESCRIPTION
**Pull request by mdickinson**
Python 3.12 is imminent. This is an experimental PR to add Python 3.12 support:

- Add 3.12 to the trove classifiers
- Add 3.12 to CI test matrices
- Allow running tests with PySide6 on Python 3.11, but not on Python 3.12
- Update cibuildwheel to latest version so that Python 3.12 eggs are built when making a release

Making this as a draft PR for now, since the possible failure modes are many and varied.

### CODE REVIEW

File: .github/workflows/run-traits-tests.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ on:
 - workflow_dispatch
 
 env:
-  # Temporary workaround prior to release of Traits 8.0
+  # Temporary workaround prior to release of TraitsUI 8.0
```
Comment (mdickinson) [inline .github/workflows/run-traits-tests.yml:8]: Drive-by fix while we're touching this file, unrelated to the rest of the PR.

File: .github/workflows/run-core-traits-tests.yml [hunk 2]
```diff
@@ -21,6 +21,7 @@ jobs:
       uses: actions/setup-python@v4
       with:
         python-version: ${{ matrix.python-version }}
+        allow-prereleases: true
```
Comment (mdickinson) [inline .github/workflows/run-core-traits-tests.yml:24]: This allows setup-python to install a Python 3.12 pre-release, and avoids the need to change these workflows _again_ once Python 3.12.0 is released.

File: pyproject.toml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [build-system]
-requires = ['setuptools<65.2', 'wheel']
+requires = ['setuptools', 'wheel']
```
Comment (mdickinson) [inline pyproject.toml:2]: Older versions of setuptools break with Python 3.12, so it was necessary to remove this restriction.

File: pyproject.toml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [build-system]
-requires = ['setuptools<65.2', 'wheel']
```
Comment (rahulporuri) [inline pyproject.toml:2]: I should be able to verify that the original issue (https://github.com/enthought/traits/pull/1722) got resolved by simply creating an edm environment and pip installing traits locally?

**Review by rahulporuri — APPROVED** [review]
LGTM with one minor question/clarification

File: pyproject.toml (hunk 4 above)
Comment (mdickinson) [reply pyproject.toml:2]: Sorry, I missed this. Yes, that should work.

Comment (rahulporuri) [reply pyproject.toml:2]: yep yep it worked for me locally on python 3.11.4 edm
